core/src/main/scala/kafka/server/LogOffsetMetadata.scala

```diff
@@ -58,8 +58,9 @@ case class LogOffsetMetadata(messageOffset: Long,
     this.segmentBaseOffset == that.segmentBaseOffset
   }
 
-  // check if this offset is before the given offset
-  def precedes(that: LogOffsetMetadata): Boolean = this.messageOffset < that.messageOffset
+  // check if this offset is before the given offset. We need to compare both message offset and segment base offset.
+  def precedes(that: LogOffsetMetadata): Boolean = this.messageOffset < that.messageOffset ||
+    (this.messageOffset == that.messageOffset && this.segmentBaseOffset < that.segmentBaseOffset)
```

Comment:
Thanks for the explanation @becketqin. I agree that it would be clearer if we check that too. Also worth keeping in mind that every other method that accesses `segmentBaseOffset` throws an exception if `messageOffsetOnly()` is true. It would be good to remain consistent, if it makes sense.